### ISSUE TITLE
monitor: Reduce overhead

### ISSUE THREAD

**Issue by aalemayhu** [cilium/cilium#1926]
We are encoding / decoding a lot. I think this is causing the node monitor to show up in perf and causes the CPU spikes. Could we recycle some of the arrays and shrink / grow them instead of creating new ones? 

- [ ] TODO: add measurments

Options:
 - [x] Use faster JSON library instead of gob (https://journal.paul.querna.org/articles/2014/03/31/ffjson-faster-json-in-go/)
 - [x] Implement server side filter to avoid unnecessary messages to clients

**Comment by raybejjani** [comment]
Some thing that might be worth investigating:
- How fast is gob to begin with?
- Can gob use a memory pool to avoid object churn
- is the slice make on pkg/bpf/perf.go:306 on-stack or heap? Nominally the escape analysis should put it on stack but we take an address of it and we pass it to `unsafe` and that might cause it to go on-heap.
- We malloc a buffer on pkg/bpf/perf.go:307 and free it in every read, this is probably not too bad but since we know we have only one perf reader, it might be possible to reuse it.
- Since we're using cgo there might be some scheduling quirks. When a gorouting enters a c context the thread it is in is blocked on C and this might make go spawn new threads. It may also get pegged to a cpu but I'm not too sure.

**Comment by aalemayhu** [comment]
@raybejjani had some great ideas shared on a call. We can reduce the state allocation on read calls by moving it into the perf structure. To measure changes we can use [ReadMemStats](https://golang.org/pkg/runtime/#ReadMemStats) before and after. Would be nice to be able to trigger this via signal. Will work on this next.

Thanks.

**Comment by raybejjani** [comment]
Ok. here are some numbers from the `bench_gob` branch. I've tried to break down the various steps we take. The Write* functions allocate a buffer and a gob encoder. In trying to optimise all of it out, I encountered an odd gob behaviour: Type information is only written out the first time gob encodes something. This causes errors on the reader side.
```
BenchmarkWriteMetaPayload-2       300000              5058 ns/op            1512 B/op         29 allocs/op
BenchmarkReadMetaPayload-2         50000             29141 ns/op           12312 B/op        212 allocs/op
BenchmarkEncodeMetaPayload-2      300000              4096 ns/op            1456 B/op         25 allocs/op
BenchmarkDecodeMetaPayload-2       50000             27439 ns/op            8032 B/op        205 allocs/op
BenchmarkMetaGobEncode-2         2000000               920 ns/op             168 B/op          5 allocs/op
BenchmarkMetaGobDecode-2           50000             23303 ns/op            7384 B/op        185 allocs/op
BenchmarkPayloadGobEncode-2      3000000               560 ns/op               0 B/op          0 allocs/op
BenchmarkPayloadGobDecode-2        50000             26863 ns/op            7920 B/op        204 allocs/op
```
The `bench_gob` branch reuses buffers and I'll try to run it under some load to get a sense if that's improved memory footprint.

**Comment by raybejjani** [comment]
Hmm. I tried ffjson out. It is definitely easier to use! It also natively supports using a buffer pool or otherwise reusing buffers. It does require some code generation, however.
It's faster than gob in decoding and uses less memory. It's slower on the encode, however, which is somewhat expected given the format.
```
BenchmarkPayloadJsonEncode-2     2000000               703 ns/op            1248 B/op          4 allocs/op
BenchmarkPayloadJsonDecode-2     2000000               679 ns/op               8 B/op          1 allocs/op
```

**Comment by raybejjani** [comment]
I ran a small experiment while waiting for VMs to build: https://play.golang.org/p/y0MEPJ4bMG. It tries to encode our Payload struct to compare how big the data is with and without the gob type data. It is 61 bytes vs 4 for an empty copy of a Payload!
We currently create a new gob encoder per packet, and encode using that. gob will send type data when it first encodes a type, and the decoder expects this on the other end. As I discovered, reusing a gob encoder without a matching decoder will break things. We can create a gob encoder for each reader in the node-monitor, and the reader already creates a gob decoder. Reusing this encoder should be safe and will shave of 57 bytes per packet, and avoid some allocations on the monitor side (the gob.NewEncoder allocates some slices and dicts).

I'll make a PR of this tomorrow.

**Comment by aalemayhu** [comment]
@raybejjani did you open a pull request for this? Couldn't find it.

**Comment by raybejjani** [comment]
@scanf I'm afraid I didn't :/ I lost a few days trying to fix a crash bug and this because a medium priority ticket so I'm trying to get my 1.0 high priority ones done first :/ I just pushed up my bench_gob branch which has the benchmarks I've been running if you want to have a look.
Still, I'm confident that we'd reduce the amount of data we're moving about by having an encoder per reader connection but we'd need to make sure all readers switch to the new scheme too (they must currently also be creating a new decoder for every read... perhaps even two since we write Meta and Payload with separate encoders currently).

**Comment by robermorales** [comment]
Some news?

**Comment by ianvernon** [comment]
@robermorales we have made several enhancements on the monitor side, including the following to reduce overhead:

* periodic aggregation of datapath notifications: 
https://github.com/cilium/cilium/pull/4936
* more efficient monitor event coding: https://github.com/cilium/cilium/pull/4581

These are upcoming in v1.2 in Cilium.

**Comment by tgraf** [comment]
We have merged all original ideas on, this closed by #4936 and #4581

**Comment by robermorales** [comment]
Cool. Thanks.
